src/Symfony/Component/Serializer/Instantiator/Instantiator.php

```diff
@@ -0,0 +1,121 @@
+
+use Symfony\Component\PropertyInfo\PropertyListExtractorInterface;
+use Symfony\Component\Serializer\Exception\MissingConstructorArgumentsException;
+use Symfony\Component\Serializer\Exception\RuntimeException;
+use Symfony\Component\Serializer\NameConverter\NameConverterInterface;
+use Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer;
+use Symfony\Component\Serializer\Normalizer\DenormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\DenormalizerAwareTrait;
+use Symfony\Component\Serializer\Normalizer\ObjectToPopulateTrait;
+
+/**
+ * @author Jérôme Desjardins <jewome62@gmail.com>
+ */
+class Instantiator implements InstantiatorInterface, DenormalizerAwareInterface
+{
+    public const DEFAULT_CONSTRUCTOR_ARGUMENTS = AbstractObjectNormalizer::DEFAULT_CONSTRUCTOR_ARGUMENTS;
+
+    use ObjectToPopulateTrait;
+    use DenormalizerAwareTrait;
+
+    private $classDiscriminatorResolver;
+    private $propertyListExtractor;
+    private $nameConverter;
+
+    public function __construct(PropertyListExtractorInterface $propertyListExtractor = null, NameConverterInterface $nameConverter = null)
+    {
+        $this->propertyListExtractor = $propertyListExtractor;
+        $this->nameConverter = $nameConverter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function instantiate(string $class, $data, $format = null, array $context = [])
+    {
+        $allowedAttributes = $this->propertyListExtractor ? $this->propertyListExtractor->getProperties($class, $context) : null;
+        $reflectionClass = new \ReflectionClass($class);
+        $constructor = $reflectionClass->getConstructor();
+
+        if ($constructor) {
+            $constructorParameters = $constructor->getParameters();
+
+            $params = [];
+            foreach ($constructorParameters as $constructorParameter) {
+                $paramName = $constructorParameter->name;
+                $key = $this->nameConverter ? $this->nameConverter->normalize($paramName, $class, $format, $context) : $paramName;
+                $allowed = null === $allowedAttributes || \in_array($paramName, $allowedAttributes, true);
+
+                if ($allowed && $constructorParameter->isVariadic() && \array_key_exists($key, $data)) {
+                    if (!\is_array($data[$paramName])) {
+                        throw new RuntimeException(sprintf('Cannot create an instance of %s from serialized data because the variadic parameter %s can only accept an array.', $class, $constructorParameter->name));
+                    }
+
+                    $params = array_merge($params, $data[$paramName]);
+                } elseif ($allowed && \array_key_exists($key, $data)) {
+                    $parameterData = $data[$key];
+                    if (null === $parameterData && $constructorParameter->allowsNull()) {
+                        $params[] = null;
+                        // Don't run set for a parameter passed to the constructor
```

Comment:
this won't work as you never return the modified `$data` outside the class with this extracted implementation.

Reply:
You are right, i remove this comment and the unset part ATM, i have 2 ideas to fix this, but it should not be done here:

 * We could pass $data as reference (not a fan)
 * We could add a filter on properties list that remove properties set with constructor (would be the best)